### PR TITLE
Prevent `</p>` output in Rich Text control

### LINKED ISSUE #437: Empty Rich Text field outputs </p>
Hi, if you add a Rich Text field to a block but don't enter any text then it outputs a `</p>` tag in the frontend.  The expected behaviour would be a blank string similar to how the Text field functions.  This happens when the RTE is a top level field and also when it is added as a sub field to a repeater.

Black Lab 1.4.1

### PR DESCRIPTION
**Pull request by kienstra**
## Steps to reproduce
1. Create a new post
2. Add a Block Lab block that has a Rich Text field without a default value
3. Don't enter any text in the Rich Text field
4. Publish the post, and go to the front-end
5. Expected: The output is well-formed html
6. Actual: The output is `</p>`, as reported in #437:

closing-p

It looks like when the <RichText> component has no text, it saves `<p></p>`. But passing that to wpautop() returns `</p>`:

wpautop-here

So in that case, this PR simply returns the string.

Closes #437

### CODE REVIEW

**Comment by kienstra** [comment]
**Request For Review**

Hi @lukecarbis,
When you have a chance, could you please review this? 

Thanks!

File: php/blocks/controls/class-rich-text.php [hunk 1]
```diff
@@ -51,6 +51,12 @@ public function register_settings() {
 	 */
 	public function validate( $value, $echo ) {
 		unset( $echo );
+
+		// For this value, wpautop() returns '</p>', so simply return the value.
+		if ( '<p></p>' === $value ) {
```
Comment (lukecarbis) [inline php/blocks/controls/class-rich-text.php:56]: Something like this might handle more weird edge cases (such as `<span></span>` or `<p><strong></strong></p>`):

```PHP
if ( empty( esc_html( $value ) ) ) {
    return $value;
}
// or…
if ( '' === esc_html( $value ) ) {
    return $value;
}
// or even…
if ( '' === esc_html( $value ) ) {
    return '';
}
```

**Review by lukecarbis — APPROVED** [review]
One small suggestion here, feel free to merge with or without it.

File: php/blocks/controls/class-rich-text.php (hunk 1 above)
Comment (kienstra) [reply php/blocks/controls/class-rich-text.php:56]: Ah, interesting point.

I'm not sure if there will be a case where `empty( esc_html( $value ) )`, or `'' === esc_html( $value )`, though maybe I just haven't seen it. 

It looks like even if there's no value entered in the Rich Text control, its value is `<p></p>`

Comment (lukecarbis) [reply php/blocks/controls/class-rich-text.php:56]: I could be missing something here.
- If the Rich Text control value is empty, the value is set to `<p></p>`
- The expected value would be an empty string, right?
- The logic above says "After removing all the HTML, if I'm left with an empty string, return an empty string"

Comment (kienstra) [reply php/blocks/controls/class-rich-text.php:56]: Hi @lukecarbis,
All good, those are good points.

>If the Rich Text control value is empty, the value is set to `<p></p>`

Right 😄 

>The expected value would be an empty string, right?

Good question, maybe 😄 

The way I thought of this was like the Paragraph block, which also uses the `<RichText>` component. It outputs `<p></p>` when it's empty, just like this Rich Text control does:

empty-paragraph-block

This is definitely debatable, though. The paragraph block isn't exactly like this.


>The logic above says "After removing all the HTML, if I'm left with an empty string, return an empty string"

Ah, `esc_html()` doesn't usually remove HTML, it just escapes it. 

For example, `esc_html( '<div></div>' )` returns `&lt;div&gt;&lt;/div&gt;`

Comment (kienstra) [reply php/blocks/controls/class-rich-text.php:56]: Maybe [strip_tags()](https://www.php.net/manual/en/function.strip-tags.php)

Comment (lukecarbis) [reply php/blocks/controls/class-rich-text.php:56]: Ah – of course! Sorry, I did mean strip_tags. Even better, though: [`wp_strip_all_tags`](https://developer.wordpress.org/reference/functions/wp_strip_all_tags/).

Comment (kienstra) [reply php/blocks/controls/class-rich-text.php:56]: Sure, I thought that's what you meant.

Comment (kienstra) [reply php/blocks/controls/class-rich-text.php:56]: Though it's unlikely, it's possible that a Rich Text field has only an image, in which case it would be stripped entirely, as there's no text other than the markup:

image-in-rich-text

Comment (kienstra) [reply php/blocks/controls/class-rich-text.php:56]: https://github.com/getblocklab/block-lab/pull/444/commits/a65299c07cbe1432f27b760aa511c2dcc12b4ef6 changes this to return `''` if the Rich Text field is empty (if it saves `<p></p>`).

**Comment by kienstra** [comment]
Hi @lukecarbis,
Happy weekend!

What do you think of the [latest changes](https://github.com/getblocklab/block-lab/pull/444#discussion_r332307316) here?

Thanks!

File: php/blocks/controls/class-rich-text.php (hunk 1 above)
Comment (lukecarbis) [reply php/blocks/controls/class-rich-text.php:56]: Aah! Great pickup on the image!

**Comment by lukecarbis** [comment]
Great! Going to merge this on you behalf, if you don't mind, so we can get that unit test fix in a couple of other PRs.

File: php/blocks/controls/class-rich-text.php (hunk 1 above)
Comment (kienstra) [reply php/blocks/controls/class-rich-text.php:56]: Thanks, it's kind of an edge case. Otherwise, `strip_tags()` would work really well.

**Comment by kienstra** [comment]
Sounds good, thanks for merging this.
